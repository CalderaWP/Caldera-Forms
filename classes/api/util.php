<?php


/**
 * Utility functions for Caldera Forms REST API
 *
 * @package Caldera_Forms
 * @author    Josh Pollock <Josh@CalderaWP.com>
 * @license   GPL-2.0+
 * @link
 * @copyright 2016 CalderaWP LLC
 */
class Caldera_Forms_API_Util {

	/**
	 * The namespace for Caldera Forms REST API
	 *
	 * @since 1.4.4
	 *
     * @param string $version Optional. API version. Default is v2. Can be v3 or v2 only
	 * @return string
	 */
	public static function api_namespace($version = 'v2'){
        $version = in_array( $version, ['v2', 'v3' ] ) ? $version : 'v2';
		return "cf-api/$version";
	}

	/**
	 * The URL for Caldera Forms REST API
	 *
	 * @since 1.4.4
	 *
	 * @param string $endpoint Optional. Endpoint.
	 * @param bool $add_nonce Optional. If true, _wp_nonce is set with WP REST API nonce. Default is false
	 * @param string $version Optional. @since 1.8.0
	 * @return string
	 */
	public static function url( $endpoint = '', $add_nonce = false, $version = 'v2' ){
		if( ! function_exists( 'rest_url' ) ){
			return '';
		}

		$url =  rest_url( self::api_namespace($version) . '/' . $endpoint );
		if( $add_nonce ){
			$url = add_query_arg( '_wpnonce', self::get_core_nonce(), $url );
		}

		return $url;
	}

	public static function check_api_token( WP_REST_Request $request ){
		$allowed = false;
			if( false != ( $token = $request->get_header( 'x_cf_entry_token') ) && is_string( $request[ 'form_id' ] ) ){
				$allowed = Caldera_Forms_API_Token::check_token( $token, $request[ 'form_id' ] );
			}

		return $allowed;
	}

	/**
	 * Get the nonce for the WP REST API
	 *
	 * @since 1.5.3
	 *
	 * @return string
	 */
	public static function get_core_nonce(){
		return wp_create_nonce( 'wp_rest' );
	}

    /**
     * Given an array of field IDs and form config, reduce field IDs array to fields that form has
     *
     * @since 1.7.0
     *
     * @param array $field_ids
     * @param array $form
     * @return array
     */
	public static function validate_array_of_field_ids(array $field_ids, array $form )
    {
        $valid_fields = [];
        $form_fields = array_keys(Caldera_Forms_Forms::get_fields( $form, false ));
        if (! empty( $field_ids )) {
            foreach ($field_ids as $field_id) {
                $field_id = trim( $field_id );
                if( in_array( $field_id, $form_fields ) ){
                    $valid_fields[] = $field_id;
                }

            }
        }

        return $valid_fields;
    }

}