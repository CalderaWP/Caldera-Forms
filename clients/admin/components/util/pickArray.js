export const pickArray = (array, key) => {
	return array.reduce(
		(accumualtor, item) =>
			accumualtor.concat([item[key]]), []
	);
}