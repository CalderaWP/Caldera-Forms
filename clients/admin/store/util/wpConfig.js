/**
 * API config and other stuff passed to jsLand via wp_localize_script
 */
export default {
    cfdotcom: 'object' === typeof CF_CLIPPY ? CF_CLIPPY.cfdotcom : {},
    fallback: 'object' === typeof CF_CLIPPY ? CF_CLIPPY.fallback : {},
    extendTitle:  'object' === typeof CF_CLIPPY ? CF_CLIPPY.extend_title : '',
    noForms:  'object' === typeof CF_CLIPPY ? CF_CLIPPY.noForms : '',
};
