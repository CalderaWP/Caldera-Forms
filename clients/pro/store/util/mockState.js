export const mockSTATE = {
	loading: false,
	connected: true,
	forms: [
		{ form_id : 1, name: 'One', layout: 2, pdf_layout:1 },
		{ form_id : 2, name: 'Two', layout: 1, pdf_layout:2 },
	],
	settings : {
		enhancedDelivery: true,
		generatePDFs: false
	},
	layouts : [
		{ id: 1, name: 'One' },
		{ id: 2, name: 'Two' }
	],
	account: {
		plan: 'apex',
		id: 42,
		apiKeys: {
			public: 'public',
			secret: 'secret',
			token: 'token'
		}
	}

};