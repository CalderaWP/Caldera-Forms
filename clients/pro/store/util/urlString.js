export  const urlString = function (data, endpoint) {
	var str = "";
	for (var key in data) {
		if (str != "") {
			str += "&";
		}
		str += key + "=" + data[key];
	}
	return endpoint + '?' + str;
};
