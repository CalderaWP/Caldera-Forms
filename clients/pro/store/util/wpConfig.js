export default {
	/**
	 * The URL for local site API
	 *
	 * @since 0.0.1
	 *
	 * @type {string}
	 */
	localApiURL: CF_PRO_ADMIN.api.cf.url,

	/**
	 * The nonce for local site API
	 *
	 * @since 0.0.1
	 *
	 * @type {string}
	 */
	localApiNonce: CF_PRO_ADMIN.api.cf.nonce,

	/**
	 * The URL for remote app API
	 *
	 * @since 0.0.1
	 *
	 * @type {string}
	 */
	appURL: CF_PRO_ADMIN.api.cfPro.url,

	/**
	 * The ID of the form whose edit screen we are on
	 *
	 * @since 1.0.0
	 *
	 * @type {string}
	 */
	formScreen: CF_PRO_ADMIN.formScreen,

	/**
	 * Translation and other strings.
	 *
	 * Created by "strings" index of generated in calderawp\calderaforms\pro\admin\scripts::data()
	 *
	 * @since 1.0.0
	 *
	 * @type {Object}
	 */
	strings: CF_PRO_ADMIN.strings,

    /**
     * Levels of Monolog logs
     *
     * Created by "logLevels" index of generated in calderawp\calderaforms\pro\admin\scripts::data()
     *
     * @since 1.5.9
     *
     * @type array
     */
    logLevels: CF_PRO_ADMIN.logLevels,

};
