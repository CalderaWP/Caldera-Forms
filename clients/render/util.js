const SparkMD5 = require('spark-md5');

export const getFieldConfigBy = (fieldConfigs, findBy, findWhere) => {
	return fieldConfigs.find(field => findWhere === field[findBy]);

};


export const hashFile = (file, callback) => {
	let blobSlice = File.prototype.slice || File.prototype.mozSlice || File.prototype.webkitSlice,
		chunkSize = 2097152,                             // Read in chunks of 2MB
		chunks = Math.ceil(file.size / chunkSize),
		currentChunk = 0,
		spark = new SparkMD5.ArrayBuffer(),
		fileReader = new FileReader();

	fileReader.onload = function (e) {
		spark.append(e.target.result);                   // Append array buffer
		currentChunk++;

		if (currentChunk < chunks) {
			loadNext();
		} else {
			callback(spark.end());
		}
	};

	fileReader.onerror = function () {
		console.warn('oops, something went wrong.');
	};

	function loadNext() {
		var start = currentChunk * chunkSize,
			end = ((start + chunkSize) >= file.size) ? file.size : start + chunkSize;

		fileReader.readAsArrayBuffer(blobSlice.call(file, start, end));
	}

	loadNext();
}



export const removeFromPending = (fieldId, cf2) => {
	const index = cf2.pending.findIndex(item => item === fieldId);
	if (-1 < index) {
		cf2.pending.splice(index, 1);
	}
}

export const removeFromUploadStarted = (fieldId, cf2) => {
	const index = cf2.uploadStarted.findIndex(item => item === fieldId);
	if (-1 < index) {
		cf2.uploadStarted.splice(index, 1);
	}
}

export const removeFromBlocking = (fieldId, cf2, fieldConfig = {}) => {
	const index = cf2.fieldsBlocking.findIndex(item => item === fieldId);
	if (-1 < index) {
		cf2.fieldsBlocking.splice(index, 1);
	}
	setSubmitButtonState(cf2, fieldConfig);
}

export const setBlocking = ( fieldId, cf2, fieldConfig = {}) => {
	removeFromUploadStarted(fieldId, cf2);
	removeFromPending(fieldId, cf2);
	if(cf2.fieldsBlocking.indexOf(fieldId) < 0){
		cf2.fieldsBlocking.push( fieldId );
	}
	setSubmitButtonState(cf2, fieldConfig, false);
}

/**
 * Default process to convert a file to media
 *
 * @param file
 * @param additionalData
 * @param fetch
 * @return {*}
 */
export const createMediaFromFile = (file, additionalData, fetch) => {

	// Create upload payload
	const data = new window.FormData();
	data.append('file', file, file.name || file.type.replace('/', '.'));
	data.append('title', file.name ? file.name.replace(/\.[^.]+$/, '') : file.type.replace('/', '.'));
	Object.keys(additionalData)
		.forEach(key => data.append(key, additionalData[key]));

	return fetch(additionalData.API_FOR_FILES_URL, {
		body: data,
		method: 'POST',
		headers: {
			'X-WP-Nonce': additionalData._wp_nonce
		}
	});
}

/**

 * Converts the bytes values to smaller value in corresponding unit
 *
 * @param bytes value to be converted
 * @return {*}
 */
export const sizeFormat = (bytes) => {
	let converted = false;
	const quant = [
		{
			unit: 'TB',
			mag: 1099511627776
		},

		{
			unit: 'GB',
			mag: 1073741824
		},

		{
			unit: 'MB',
			mag: 1048576
		},

		{
			unit: 'kB',
			mag: 1024
		},

		{
			unit: 'B ',
			mag: 1
		}
	];
	quant.forEach(function(v){
		if (parseFloat(bytes) >= v.mag && converted == false){
			converted = bytes/v.mag;
			if( bytes > 1048576 ){
				converted = converted.toFixed(2);
			}else{
				converted = Math.round( converted );
			}
			converted = converted +' '+v.unit;
		}
	});
	return converted;
}

/* Set Submit button state
 *
 * @param {object} cf2
 * @param {object} fieldConfig
 * @param {boolean} state
 * @return {*}
 */
export const setSubmitButtonState = (cf2, fieldConfig, state) => {

	const fieldIdAttr = fieldConfig.fieldIdAttr;

	const formIdAttr = getFormIdAttr(cf2, fieldIdAttr);

	const form =  jQuery("#" + formIdAttr);
	//If no state param was send in the function define the state based on elements inside the fieldsBlocking Array
	if(typeof state === "undefined") {
		state = cf2.fieldsBlocking.length <= 0;
	}

	//If state === true enable submit button else disable submit button
	if(state) {
		form.find(':submit').prop('disabled',false);
	} else {
		form.find(':submit').prop('disabled',true);
	}
	return state;
}

/**
 * Find the formIdAttr corresponding to a fieldIdAttr in the cf2 object
 *
 * @param {object} cf2
 * @param {string} fieldIdAttr
 * @return {string} formIdAttr
 */
export const getFormIdAttr = (cf2, fieldIdAttr) => {

	let formIdAttr = cf2.formIdAttr;

	if( typeof formIdAttr === "undefined" ){

		const entries = Object.entries(cf2);
		const formEntries = [];
		entries.forEach( entry => {
			if(entry[1].hasOwnProperty("fields") ) {
				formEntries.push(entry);
			}
		})
		formEntries.forEach( formEntry => {
			if(formEntry[1].fields.hasOwnProperty(fieldIdAttr ) ) {
				formIdAttr = formEntry[0];
			}
		})
	}

	return formIdAttr;

}

/**
 * Start or Stop spinner animation to indicate process is happening
 *
 * @since 1.8.0
 *
 * @param {string} formIdAttr Id Attribute of the <form> tag
 * @param {boolean} trigger if true force addClass if false force removeClass
 */
export const processAnimation = (formIdAttr, trigger) => {

	const cfElement = formIdAttr.length > 0 ? jQuery("#" + formIdAttr).parent(".caldera-grid") : false;

	if(cfElement !== false) {
		if(trigger === false){
			if(jQuery(cfElement).hasClass("cf_processing")){
				jQuery(cfElement).removeClass("cf_processing");
			}
		}else if(trigger === true){
			if(!jQuery(cfElement).hasClass("cf_processing")){
				jQuery(cfElement).addClass("cf_processing");
			}
		}
	}

};

/**
 * Add a component -- CalderaFormsRender -- to window.cf2 correctly
 *
 * @since 1.8.0
 *
 * @param {function} component
 * @param {string}idAttr
 * @param {object}window
 */
export const captureRenderComponentRef = (component,idAttr,window = {}) => {
	window.cf2 = window.cf2 || {};
	window.cf2[idAttr] = {
		...window.cf2[idAttr] || {},
		component
	};
};
