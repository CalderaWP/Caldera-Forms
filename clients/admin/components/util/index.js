import {optionFactory} from "./optionFactory";
import {pickArray} from "./pickArray";
import {settingsProp} from "../../state/prepareSettings";

export default {
	optionFactory,
	pickArray,
	settingsProp
}