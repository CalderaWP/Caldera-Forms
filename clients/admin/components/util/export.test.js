import util from './index';
import {snapshotObjectKeysAndTypes} from "../../testUtil/snapshotObjectKeysAndTypes";

describe( 'util  export', () => {
	it( 'index has the right types and keys', () => {
		snapshotObjectKeysAndTypes(util);
	});
});