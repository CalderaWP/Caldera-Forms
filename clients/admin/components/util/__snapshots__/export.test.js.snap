// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util  export index has the right types and keys 1`] = `"{\\"optionFactory\\":\\"function\\",\\"pickArray\\":\\"function\\",\\"settingsProp\\":\\"function\\"}"`;
